File: dc-02-19-19/producer/src/test/java/com/example/demo/DemoApplicationTests.java
package com.example.demo;


import org.hamcrest.Matcher;
import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.reactivestreams.Publisher;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.data.mongo.DataMongoTest;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;


@DataMongoTest
public class DemoApplicationTests {

    @Autowired
    PersonRepository repo;

    Matcher<Person> personMatcher = Matchers.allOf(Matchers.hasProperty("name", Matchers.equalTo("Bob")),
            Matchers.hasProperty("id"), Matchers.notNullValue()
    );

    @Test
    public void testShouldAssertEntity() {
        Person p = new Person("1234", "Bob");

        Assertions.assertNotNull(p);
        Assertions.assertNotSame(p, new Person("1234", "Bob"));

        MatcherAssert.assertThat("Person has normal state",
                p,
                personMatcher);
    }

    @Test
    public void testShouldSaveFind() {
        Person p = new Person(null, "Bob");

        Publisher<Person> save = repo.save(p);
        Publisher<Person> find = repo.findAll();
        Publisher<Person> composition = Flux.from(save)
                                        .thenMany(find);

        StepVerifier
                .create(composition)
                .expectSubscription()
                .assertNext(it -> MatcherAssert.assertThat("state", it, personMatcher))
                .expectComplete()
                .verify();
    }


}

File: chi-07-10-19/producer/src/test/kotlin/com/example/producer/DemoApplicationTests.kt
package com.example.producer

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.data.mongo.DataMongoTest
import org.springframework.data.mongodb.repository.ReactiveMongoRepository
import reactor.test.StepVerifier
import java.util.*

@DataMongoTest
class DemoApplicationTests {

    @Autowired
    private lateinit var repo: ReactiveMongoRepository<Message, UUID>

    @Test
    fun `Should save and load Entities`() {
        val message = Message(UUID.randomUUID(), "Hello")

        StepVerifier
                .create(repo
                        .save(message)
                        .thenMany(repo.findAll()))
                .expectSubscription()
                .assertNext {
                    Assertions
                            .assertThat(it)
                            .isNotNull
                            .hasNoNullFieldsOrProperties()
                            .hasFieldOrPropertyWithValue("text", "Hello")
                }
                .expectComplete()
                .verify()
    }

}
